replica/database.hh

```diff
@@ -351,6 +385,8 @@ struct table_stats {
     int64_t memtable_range_tombstone_reads = 0;
     int64_t memtable_row_tombstone_reads = 0;
     int64_t tablet_count = 0;
+    consumption_unit_counter consumed_read_units;
+    consumption_unit_counter consumed_write_units;
```

Comment:
When a table is removed, need to move the counter to the database to a counter of all closed tables.

Need to somehow persist on shutdown.

Reply:
We will start by relaying on an external metrics storage

Reply:
We should have designed this project before doing it - then you could have seen how this PR fits in the bigger picture.

Basically, the metric you see here is just a normal Scylla metric - like we have "number of CQL requests" we now also have a pretty "consumed write units" graph to look at. Like "number of CQL requests" this one is also per-node and when a node reboots it begins at zero again and when a node is decommissioned, its count is no longer added to the total. So such unreliable metrics are NOT directly usable to bill users in PAY_PER_REQUEST billing mode and to do that we'll need another layer of logging these metrics in a persistent manner.

But this metric **can** already be used directly to implement PROVISIONED billing mode. In this billing mode you would have a provisioned throughput - e.g., 1000 RCU per second. Each shard can use its consumed_read_units counter (which will probably become more than a counter, also including some sort of bucket or counter-in-last-minute or something) to see if it did "too many" requests on average recently and if it did, fail the next request. To do this, persistence on shutdown or node decomission and so on is not needed.

Probably none of the above ideas will end up in the open-source version. The open-source version apparently (according to this PR) will have a nice metric graph and a per-request RCU/WCU estimate (part of the DynamoDB API) and that's it.

Reply:
Either all of this should be in open-source, or none. I can't review something like that.

Reply:
All the code here is open source and functional. Future commits may be in enterprise only, but the separation between enterprise and open source will be in specific files, not in our core files, which are frequently changing and are likely to get conflicts.